### PR TITLE
fixed Empire CheckProductionProgress so that if two identical …

### PR DESCRIPTION
**Pull request by Dilvish-fo**
…self-blocking items (such as a terraforming building) complete on the same turn at the same location, the second one is halted instead of being entirely erased.

It seems like the forum discussion favored a solution like this...

### CODE REVIEW

File: Empire/Empire.cpp [hunk 1]
```diff
@@ -2800,6 +2800,42 @@ void Empire::CheckProductionProgress() {
...
+        if (!system) {
+            ErrorLogger() << "Empire::CheckProductionProgress couldn't get system for producing new " << build_description;
+            continue;
+        }
+
+        // check location condition before each item is created, so
+        // that items being produced can prevent subsequent
+        // completions on the same turn from going through
+        if (!this->ProducibleItem(elem.item, elem.location)) {
+                    DebugLogger() << "Location test failed for " << build_description << " at location " << build_location->Name();
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: weird indenting?

File: Empire/Empire.cpp [hunk 2]
```diff
@@ -2800,6 +2800,42 @@ void Empire::CheckProductionProgress() {
...
+            ErrorLogger() << "Empire::CheckProductionProgress couldn't get system for producing new " << build_description;
+            continue;
+        }
+
+        // check location condition before each item is created, so
+        // that items being produced can prevent subsequent
+        // completions on the same turn from going through
+        if (!this->ProducibleItem(elem.item, elem.location)) {
+            DebugLogger() << "Location test failed for " << build_description << " at location " << build_location->Name();
+            break;
```
Comment (geoffthemedio) [inline Empire/Empire.cpp]: The code you moved to here was in a switch block, but this code is not, so I think this should be continue, not break, as break here will exit the for loop over queue contents.

**Comment by geoffthemedio** [comment]
I'm a bit concerned that leaving stuff on the queue when it becomes not producible, and retaining its accumulated production allocation, could allow micromanagement by stuffing the queue with a bunch of projects, manipulating their order so none finish and make all the others invalid at their current location. To avoid this, adding EnqueueLocation conditions for most stuff seems important...

File: Empire/Empire.cpp (hunk 2 above)
Comment (Dilvish-fo) [inline Empire/Empire.cpp]: ack, quite right; I'm afraid my testing had focused on the two buildings completing together, didn't pay enough attention to items later in the queue.  It's fixed now.

**Comment by Dilvish-fo** [comment]
Well, having it take the PP away from the player (even on the last turn) and then just throw all that away seems of more concern.   We've had some discussion of this on the forums, in particular regarding terraforming, where if you want to do multiple steps you could save a few turns (a bit less than half the turns for each extra terraforming) by putting them on the queue so they run partially in parallel, but you wouldn't save any on cost.  Or at least, that's how I had expected it would have been working, and that's how the PR winds up making it.  There is no special manipulation of order involved, so it doesn't seem like micromanagement really any more than building multiple ships in a block, which can save a turn or so compared to having them all as separate builds. 

For any future buildings where there was some more micromanagement concern, then sure, Enqueue conditions could protect against that.

**Comment by geoffthemedio** [comment]
For the case of terraforming, the micromanagement isn't so much of an issue because the building deletes itself after working for a turn.

The point of the enqueue condition is, I think, to prevent multiples of things being added when it makes no sense to do so... But for things like terraforming, I feel like it shouldn't be possible to produce them in parallel, as it is a sequential process from one planet type to the next, so it seems like it should take the full production time for each step, not just half the time after the cost increases (or no time if reverting terraforming, I suppose). But perhaps a better solution is to make the terraforming buildings just not function if the planet has been terraformed recently... with the minimum time being the production time for a terraforming building on the same planet. Could implement that with a "recently terraformed" special, which deletes itself after X turns, and which cannot be present on a planet for a terraforming building to function (but which wouldn't prevent a terraforming building from being produced there).

**Comment by Dilvish-fo** [comment]
> But perhaps a better solution is to make the terraforming buildings just not function if the planet has been terraformed recently... with the minimum time being the production time for a terraforming building on the same planet. Could implement that with a "recently terraformed" special,

That seems like a lot of unnecessary hassle and complication just to ensure terraforming takes the full time, when we can simply add an already defined enqueued macro to the script instead.  

Anyway, the decision on what to do with the terraforming content is a related but different issue which could be discussed more in the forums.   The purpose of this PR is to fix the bug where the production queue would wind up discarding something you had been allowed to build.

**Pull request merged**